### DISCUSSION TITLE
On WSL:  Running Homebrew as root is extremely dangerous...

### DISCUSSION THREAD

**Discussion by yumetsuro** [Homebrew/discussions#2771] [Linux] [answered]
Hi everyone.

Today i needed Homebrew to do some things, so i tried to install following the guides, also the one on the official site.
When i finally managed to set up and get brew thought the alternative installation 
This
`git clone https://github.com/Homebrew/brew ~/.linuxbrew/Homebrew
mkdir ~/.linuxbrew/bin
ln -s ../Homebrew/bin/brew ~/.linuxbrew/bin
eval $(~/.linuxbrew/bin/brew shellenv)`

I proceeded to use it and it gives me the same error every time: 
"Error: Running Homebrew as root is extremely dangerous and no longer supported.                                                                                                                                                            As Homebrew does not drop privileges on installation you would be giving all                                                                                                                                                               build scripts full access to your system. "

I tried to use `sudo chown -R $(whoami) $(brew --prefix)/*` and other commands but nothing worked.

I'm using Ubuntu 20.04 LTS as WSL

Thanks in advance for the help.

**Comment by carlocab** [comment]
Do you have `sudo` access on your machine? If you do, I recommend doing
```
rm -rf ~/.linuxbrew
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

**Reply by yumetsuro** [reply]
When i run the second command it gives me: 

_==> Checking for `sudo` access (which may request your password)...                                                                                                                                                                        Don't run this as root!_

**Reply by pedinil** [reply]
when i was using the root password , it was saying the Sorry, try again. and i was thinking may be my password is wrong but after
running rm -rf ~/.linuxbrew , it is working fine , i think there is bug the script

**Reply by gromgit** [reply]
There may still be bugs in the installer script, but this probably isn't one of them.

> when i was using the root password

The installer uses `sudo`, which asks for _your_ password, not root's.

**Comment by gromgit** [comment, marked as answer]
What's the output of `id`?

**Reply by yumetsuro** [reply]
This: uid=0(root) gid=0(root) groups=0(root),27(sudo),1000(docker)      

Also the current user is a sudoer

**Reply by gromgit** [reply]
That's what the installer is complaining about. It expects to populate the Homebrew hierarchy as you, not root, so you have to run it as a user _other than root_, i.e. don't run the installer in a sudo environment..

**Reply by yumetsuro** [reply]
It worked thank you!

**Reply by BloemendaalEzra** [reply]
How do you change the user? my output of id is uid=0(root) gid=0(root) groups=0(root)

**Comment by lanzoninicola** [comment]
Although it was late, I believe this solution may help others in the future. Here are the steps I took to solve the problem:

I created a new user by running the following command: 
`sudo adduser [username]`

I added the new user to the "sudo" group by running the following command: 
`usermod -aG sudo [username]`

To check if the user was successfully added to the group, I ran the following command: 
`getent group <groupname> | cut-d: -f4`

I opened a new Ubuntu session to ensure that the changes took effect.

Finally, I installed Homebrew following the official documentation.

**Reply by gromgit** [reply]
> To check if the user was successfully added to the group, I ran the following command:
> `getent group <groupname> | cut-d: -f4`

`getent` is a GNUism, better to use the POSIX-standard `id`:
```
id <username> | grep sudo
```

**Comment by dersimn** [comment]
"is extremely dangerous", wow.. let's run it on Ubuntu then where your default user has password-less sudo access on all commands.. genius.

**Reply by SMillerDev** [reply]
If you have a good way to detect that programatically we'd happily review a pull request that warns users about that too.

**Reply by dersimn** [reply]
I would, but I'd also like to implement on some kind of "I know what I'm doing"-override-switch 🤔 on certain experimental machines it wouldn't be a problem for me to run this as root, tbh.

Note for future-me:
`sudo -l` outputs a list of all sudo-able commands, grep this for something like `(ALL) NOPASSWD: ALL`.

**Reply by trallnag** [reply]
There's no point in it at least for WSL. Privilege escalation to root is always possible by running powershell from WSL and then starting a shell as root in the WSL distribution. This does not require any password

**Comment by radutomy** [comment]
This is extremely annoying. I have a lot of experiemental throw-away VMs that I'm running as root.. I just discovered that it's impossible now to install homebrew on them without adding a sudo user.

**Reply by dersimn** [reply]
See this for a workaround: https://github.com/Homebrew/install/blob/6d48890a7430dc65a8e5718071ecdcee7d2a29c1/install.sh#L365

**Comment by prakersh** [comment]
touch /run/.containerenv  was the simplest solution. 
Thanks @dersimn!
